clang/labs/lab1/sredneva_anastasiya/Deprecated_sa.cpp

```diff
@@ -0,0 +1,70 @@
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Basic/Diagnostic.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+
+using namespace clang;
+
+class DeprecatedVisitor : public RecursiveASTVisitor<DeprecatedVisitor> {
+public:
+  explicit DeprecatedVisitor(bool isWarning) : isWarning(isWarning) {}
+  bool VisitFunctionDecl(FunctionDecl *fDecl) {
+    if (fDecl->getNameInfo().getAsString().find("deprecated") !=
+        std::string::npos) {
+      DiagnosticsEngine &diagn = fDecl->getASTContext().getDiagnostics();
+      unsigned diagnID;
+      if (isWarning) {
+        diagnID = diagn.getCustomDiagID(DiagnosticsEngine::Warning,
+                                        "The function name has 'deprecated'");
+        diagn.Report(fDecl->getLocation(), diagnID)
+            << fDecl->getNameInfo().getAsString();
+      } else {
+        diagnID = diagn.getCustomDiagID(DiagnosticsEngine::Error,
+                                        "The function name has 'deprecated'");
+        diagn.Report(fDecl->getLocation(), diagnID)
+            << fDecl->getNameInfo().getAsString();
+      }
```

Comment:
This is just a tip. It is your choice to apply it or not.
```suggestion
      diagnID = diagn.getCustomDiagID(
          isWarning ? DiagnosticsEngine::Warning : DiagnosticsEngine::Error,
          "The function name has 'deprecated'");
      diagn.Report(fDecl->getLocation(), diagnID)
          << fDecl->getNameInfo().getAsString();
```